### PR TITLE
Reset loss to zero on logging in Trainer to avoid bfloat16 issues

### PR DESCRIPTION
**Pull request by bminixhofer**
# What does this PR do?

I have had a weird issue with the logged loss going to zero after some training steps training with `bfloat16` on v3 TPUs:

W B Chart 11_16_2020, 12_00_44 PM

while it works correctly using 32bit precision:

W B Chart 11_16_2020, 12_01_54 PM

After some investigation I found that the `tr_loss` variable in `Trainer.train` seems to overflow after reaching 1024 (?). 
I did not track this down more closely because it is easily fixed by making `tr_loss` a regular Python float instead of a tensor. It doesn't actually need to be a tensor as it is only ever accessed by `.item()`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to the it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

@sgugger

### CODE REVIEW

**Comment by sgugger** [comment]
No, we don't want to do `loss.item()` at each step since it slows down a lot the training on TPUs. We can cast this scalar to Float if it overflows. Not super familiar with `bfloat16` on TPUs so we may have overlooked something there (on mixed precision for GPUs, the loss is always a float).

**Comment by bminixhofer** [comment]
Oh ok, I hadn't considered that. I'll take a closer look then to check what exactly causes the overflow.

**Comment by bminixhofer** [comment]
Alright, this might be a problem with XLA? I'm not at all familiar with how TPUs and XLA work internally but here is a minimal example of the problem:

```bfloat_demo.py```
```py
import torch

import torch_xla
import torch_xla.core.xla_model as xm


device = device = xm.xla_device()

loss_tensor = torch.tensor(0.0).to(device)
loss_float = 0.0

print(f"loss_tensor is on device {loss_tensor.device} with dtype {loss_tensor.dtype}")

to_add = torch.tensor(10.0).to(device)

for i in range(10):
    for _ in range(100):
        loss_tensor += to_add
        loss_float += to_add.item()

    print(loss_tensor, loss_float)
```

Running this regularly:

```
(torch-xla-1.7) bminixhofer@gerpt2:~$ python bfloat_demo.py
0.0 is on device xla:1 with dtype torch.float32
tensor(1000., device='xla:1') 1000.0
tensor(2000., device='xla:1') 2000.0
tensor(3000., device='xla:1') 3000.0
tensor(4000., device='xla:1') 4000.0
tensor(5000., device='xla:1') 5000.0
tensor(6000., device='xla:1') 6000.0
tensor(7000., device='xla:1') 7000.0
tensor(8000., device='xla:1') 8000.0
tensor(9000., device='xla:1') 9000.0
tensor(10000., device='xla:1') 10000.0
```

and with `bfloat16`:

```
(torch-xla-1.7) bminixhofer@gerpt2:~$ XLA_USE_BF16=1 python bfloat_demo.py 
2020-11-16 14:52:12.488382: I    1663 torch_xla/csrc/tensor_util.cpp:28] Using BF16 data type for floating point values
0.0 is on device xla:1 with dtype torch.float32
tensor(904., device='xla:1') 1000.0
tensor(1704., device='xla:1') 2000.0
tensor(2960., device='xla:1') 3000.0
tensor(4096., device='xla:1') 4000.0
tensor(4096., device='xla:1') 5000.0
tensor(4096., device='xla:1') 6000.0
tensor(4096., device='xla:1') 7000.0
tensor(4096., device='xla:1') 8000.0
tensor(4096., device='xla:1') 9000.0
tensor(4096., device='xla:1') 10000.0
```

but notably the issue doesn't seem to be the magnitude at all but rather how often a value is added:

(setting `to_add = torch.tensor(0.1).to(device)`)
```
(torch-xla-1.7) bminixhofer@gerpt2:~$ XLA_USE_BF16=1 python bfloat_demo.py 
2020-11-16 14:57:57.844438: I    1860 torch_xla/csrc/tensor_util.cpp:28] Using BF16 data type for floating point values
loss_tensor is on device xla:1 with dtype torch.float32
tensor(10.0625, device='xla:1') 10.009765625
tensor(22.5000, device='xla:1') 20.01953125
tensor(32., device='xla:1') 30.029296875
tensor(32., device='xla:1') 40.0390625
tensor(32., device='xla:1') 50.048828125
tensor(32., device='xla:1') 60.05859375
tensor(32., device='xla:1') 70.068359375
tensor(32., device='xla:1') 80.078125
tensor(32., device='xla:1') 90.087890625
tensor(32., device='xla:1') 100.09765625
```

So the dtype does not seem to be the problem, the problem seems to be something along the lines of not enough operations being tracked by XLA, but as I said I really don't know the internals at all so I don't want to go off on speculation here :)
Any ideas how to proceed?

**Comment by sgugger** [comment]
And if you do:
```python
import torch

import torch_xla
import torch_xla.core.xla_model as xm


device = xm.xla_device()

loss_tensor = torch.tensor(0.0).to(device)
loss_float = 0.0

print(f"loss_tensor is on device {loss_tensor.device} with dtype {loss_tensor.dtype}")

to_add = torch.tensor(10.0).to(device)

for i in range(10):
    for _ in range(100):
        loss_tensor += to_add.float()
        loss_float += to_add.item()

    print(loss_tensor, loss_float)
```
does this solve the issue?

Trying to avoid the `.item` as it triggers a synchronization of TPUs :-)

**Comment by bminixhofer** [comment]
No, still the same :(

`bfloat_demo.py`
```python
import torch

import torch_xla
import torch_xla.core.xla_model as xm


device = xm.xla_device()

loss_tensor = torch.tensor(0.0).to(device)
loss_float = 0.0

print(f"loss_tensor is on device {loss_tensor.device} with dtype {loss_tensor.dtype}")

to_add = torch.tensor(10.0).to(device)

for i in range(10):
    for _ in range(100):
        loss_tensor += to_add.float()
        loss_float += to_add.item()

    print(loss_tensor, loss_float)

```

```
(torch-xla-1.7) bminixhofer@gerpt2:~$ XLA_USE_BF16=1 python bfloat_demo.py 
2020-11-16 16:00:35.131065: I    1197 torch_xla/csrc/tensor_util.cpp:28] Using BF16 data type for floating point values
loss_tensor is on device xla:1 with dtype torch.float32
tensor(904., device='xla:1') 1000.0
tensor(1704., device='xla:1') 2000.0
tensor(2960., device='xla:1') 3000.0
tensor(4096., device='xla:1') 4000.0
tensor(4096., device='xla:1') 5000.0
tensor(4096., device='xla:1') 6000.0
tensor(4096., device='xla:1') 7000.0
tensor(4096., device='xla:1') 8000.0
tensor(4096., device='xla:1') 9000.0
tensor(4096., device='xla:1') 10000.0
```

**Comment by sgugger** [comment]
Ok investigated a bit more by asking some TPU experts and there is no way around this. So the proper fix will be to reset the loss to 0 each time we log it (instead of summing everything from the beginning). I can work on that later this week after the changes needed for v4, or you can work on it if it interests you.

**Comment by bminixhofer** [comment]
Ok thanks for the quick followup!

> So the proper fix will be to reset the loss to 0 each time we log it (instead of summing everything from the beginning). I can work on that later this week after the changes needed for v4, or you can work on it if it interests you.

Wouldn't that still have the same issue if `logging_steps` is sufficiently large?

**Comment by bminixhofer** [comment]
I think something like this could be a solution:

```python
import torch

import torch_xla
import torch_xla.core.xla_model as xm


device = xm.xla_device()

loss = 0.0

loss_agg_steps = 52
loss_agg = torch.tensor(0.0).to(device)
zero = torch.tensor(0.0).to(device)

to_add = torch.tensor(10.0).to(device)

for i in range(10):
    for j in range(100):
        loss_agg += to_add

        if j % loss_agg_steps == 0:
            loss += loss_agg.item()
            loss_agg.copy_(zero)

    loss += loss_agg.item()
    loss_agg.copy_(zero)

    print(loss)
```

updating the loss tensor for `n` steps and syncing with a Python float loss afterwards (and always syncing before logging and after the last batch in an epoch). `n = 52` is the highest that worked in this demo but maybe a more informed decision could be taken about that value.

**Comment by sgugger** [comment]
The `agg_step` would be the `logging_step` we have as training argument, the user can then tune it to their need.

**Comment by bminixhofer** [comment]
Ok, I thought it would be good to have a second parameter as the logging step is tied to other things as well but in that case I can give implementing it the way you described earlier a shot.

**Comment by bminixhofer** [comment]
The latest commit should do the trick. Just not sure if `tr_loss -= tr_loss` is the best way to reset `tr_loss` to zero.

**Comment by sgugger** [comment]
I think you also need to do something for the final reported training loss.

**Comment by bminixhofer** [comment]
Oh right, thanks. I introduced a variable `_total_loss_scalar` which handles that (comparable to `loss` in the demo code).

**Review by sgugger — APPROVED** [review]
Very clean! And thanks for adding comments on the objects. The rename `_logging_loss_scalar` to `_total_loss_scalar` makes perfect sense.

**Review by LysandreJik — APPROVED** [review]
LGTM
